### ISSUE TITLE
crossversion: investigate timeout

### ISSUE THREAD

**Issue by jbowens** [cockroachdb/pebble#2044]
Running the cross-version metamorphic tests w/ 21.2->22.1->22.2 on my gceworker, runs took on average 6 minutes. After 24 hours, one run of the crossversion metamorphic tests hit the 1hr timeout that I set. We should investigate why and determine if this configuration is exercising some pathological behavior.

```
=== RUN   TestMetaCrossVersion
    crossversion_test.go:73: Test directory: /tmp/TestMetaCrossVersion3415907446/001
    crossversion_test.go:74: Reproduction:
          go test -v -run 'TestMetaCrossVersion' --seed 1666464515427976089 --factor 10 --version 21.2,09f3473d,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.21.2.test --version 22.1,292cba9b,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.22.1.test --version 22.2,0d8cae91,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.22.2.test

    crossversion_test.go:92: 0: 21.2,09f3473d,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.21.2.test (Mode = -rwxrwxr-x)
    crossversion_test.go:92: 1: 22.1,292cba9b,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.22.1.test (Mode = -rwxrwxr-x)
    crossversion_test.go:92: 2: 22.2,0d8cae91,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.22.2.test (Mode = -rwxrwxr-x)
    crossversion_test.go:151: Running tests with version 09f3473d with 1 initial state(s).
    crossversion_test.go:157:   Running test with version 09f3473d with initial state <empty>.
    crossversion_test.go:151: Running tests with version 292cba9b with 10 initial state(s).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=4910346648737122976-opts=standard-010(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=4910346648737122976-opts=standard-020(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=4910346648737122976-opts=standard-012(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=4910346648737122976-opts=random-001(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=4910346648737122976-opts=random-007(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=4910346648737122976-opts=standard-006(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=4910346648737122976-opts=standard-015(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=4910346648737122976-opts=random-009(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=4910346648737122976-opts=standard-018(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=4910346648737122976-opts=standard-000(<empty>).
    crossversion_test.go:151: Running tests with version 0d8cae91 with 10 initial state(s).
    crossversion_test.go:157:   Running test with version 0d8cae91 with initial state sha=292cba9b-seed=2889797296425809833-opts=random-009(sha=09f3473d-seed=4910346648737122976-opts=standard-010(<empty>)).
panic: test timed out after 1h0m0s
```

[TestMetaCrossVersion3415907446.tar.gz](https://github.com/cockroachdb/pebble/files/9847493/TestMetaCrossVersion3415907446.tar.gz)

**Comment by jbowens** [comment]
Another 1h timeout:

```
=== RUN   TestMetaCrossVersion
    crossversion_test.go:73: Test directory: /tmp/TestMetaCrossVersion2329797999/001
    crossversion_test.go:74: Reproduction:
          go test -v -run 'TestMetaCrossVersion' --seed 1666592933009699099 --factor 10 --version 21.2,09f3473d,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.21.2.test --version 22.1,292cba9b,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.22.1.test --version 22.2,0d8cae91,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.22.2.test

    crossversion_test.go:92: 0: 21.2,09f3473d,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.21.2.test (Mode = -rwxrwxr-x)
    crossversion_test.go:92: 1: 22.1,292cba9b,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.22.1.test (Mode = -rwxrwxr-x)
    crossversion_test.go:92: 2: 22.2,0d8cae91,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.22.2.test (Mode = -rwxrwxr-x)
    crossversion_test.go:151: Running tests with version 09f3473d with 1 initial state(s).
    crossversion_test.go:157:   Running test with version 09f3473d with initial state <empty>.
    crossversion_test.go:151: Running tests with version 292cba9b with 10 initial state(s).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=3721257820412848381-opts=standard-000(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=3721257820412848381-opts=standard-004(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=3721257820412848381-opts=standard-011(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=3721257820412848381-opts=random-020(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=3721257820412848381-opts=standard-001(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=3721257820412848381-opts=random-008(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=3721257820412848381-opts=random-019(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=3721257820412848381-opts=standard-020(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=3721257820412848381-opts=random-005(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=3721257820412848381-opts=random-017(<empty>).
    crossversion_test.go:151: Running tests with version 0d8cae91 with 10 initial state(s).
    crossversion_test.go:157:   Running test with version 0d8cae91 with initial state sha=292cba9b-seed=18249841236693670679-opts=standard-012(sha=09f3473d-seed=3721257820412848381-opts=random-008(<empty>)).
    crossversion_test.go:157:   Running test with version 0d8cae91 with initial state sha=292cba9b-seed=18249841236693670679-opts=random-020(sha=09f3473d-seed=3721257820412848381-opts=random-005(<empty>)).
    crossversion_test.go:157:   Running test with version 0d8cae91 with initial state sha=292cba9b-seed=18249841236693670679-opts=standard-017(sha=09f3473d-seed=3721257820412848381-opts=random-008(<empty>)).
    crossversion_test.go:157:   Running test with version 0d8cae91 with initial state sha=292cba9b-seed=18249841236693670679-opts=random-003(sha=09f3473d-seed=3721257820412848381-opts=random-019(<empty>)).
panic: test timed out after 1h0m0s

```

**Comment by itsbilal** [comment]
Been stressing those seeds and those exact versions for the last ~3-4h, each run seems to be averaging ~15m on my Mac which is about 3x the runtime of a random seed but still well within the 1h timeout. Might try this on Linux if my repro attempts are unsuccessful.

Are your test binaries built with `invariants` by any chance?

**Comment by itsbilal** [comment]
Almost exactly 24h later, got another repro, SIGABRT'd the correct process this time (it was the 22.2 run) and I got this stacktrace:

```
        SIGABRT: abort
        PC=0x1babce270 m=0 sigcode=0

        goroutine 0 [idle]:
        runtime.pthread_cond_wait(0x10508ca00, 0x10508c9c0)
                /opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/runtime/sys_darwin.go:448 +0x20
        runtime.semasleep(0xffffffffffffffff)
                /opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/runtime/os_darwin.go:66 +0x78
        runtime.notesleep(0x10508c7c8)
                /opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/runtime/lock_sema.go:181 +0xc8
        runtime.mPark(...)
                /opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/runtime/proc.go:1449
        runtime.stopm()
                /opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/runtime/proc.go:2228 +0x88
        runtime.findrunnable()
                /opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/runtime/proc.go:2804 +0x980
        runtime.schedule()
                /opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/runtime/proc.go:3187 +0x238
        runtime.park_m(0x15f0?)
                /opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/runtime/proc.go:3336 +0x13c
        runtime.mcall()
                /opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/runtime/asm_arm64.s:190 +0x54

        goroutine 1 [chan receive]:
        testing.(*T).Run(0x1400022a820, {0x104c011f4?, 0x17049ca91fa8?}, 0x104d910d8)
                /opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/testing/testing.go:1487 +0x314
        testing.runTests.func1(0x0?)
                /opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/testing/testing.go:1839 +0x74
        testing.tRunner(0x1400022a820, 0x1400020dcb8)
                /opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/testing/testing.go:1439 +0x110
        testing.runTests(0x14000145ea0?, {0x105084240, 0x14, 0x14}, {0x1e00000000000000?, 0x104742560?, 0x0?})
                /opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/testing/testing.go:1837 +0x3f0
        testing.(*M).Run(0x14000145ea0)
                /opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/testing/testing.go:1719 +0x500
        main.main()
                _testmain.go:85 +0x1d0

        goroutine 19 [sync.Cond.Wait]:
        sync.runtime_notifyListWait(0x1400042c250, 0x1)
                /opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/runtime/sema.go:513 +0x170
        sync.(*Cond).Wait(0x1400042c240)
                       /opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/sync/cond.go:56 +0xd8
        github.com/cockroachdb/pebble.(*DB).compactMarkedFilesLocked(0x1400042c000)
                /Users/bilal/go/src/github.com/cockroachdb/pebble/format_major_version.go:416 +0x34
        github.com/cockroachdb/pebble.glob..func15(0x104d07d20?)
                /Users/bilal/go/src/github.com/cockroachdb/pebble/format_major_version.go:273 +0x24
        github.com/cockroachdb/pebble.(*DB).ratchetFormatMajorVersionLocked(0x1400042c000, 0xb)
                /Users/bilal/go/src/github.com/cockroachdb/pebble/format_major_version.go:351 +0x1f8
        github.com/cockroachdb/pebble.Open({0x1400014ca20, 0x29}, 0x14000221900)
                /Users/bilal/go/src/github.com/cockroachdb/pebble/open.go:268 +0xf54
        github.com/cockroachdb/pebble/internal/metamorphic.(*test).init.func12()
                /Users/bilal/go/src/github.com/cockroachdb/pebble/internal/metamorphic/test.go:126 +0x44
        github.com/cockroachdb/pebble/internal/metamorphic.withRetries(0x140002096f8)
                /Users/bilal/go/src/github.com/cockroachdb/pebble/internal/metamorphic/retryable.go:21 +0x34
        github.com/cockroachdb/pebble/internal/metamorphic.(*test).init(0x140002300e0, 0x140003dbd20, {0x1400014ca20, 0x29}, 0x0?)
                /Users/bilal/go/src/github.com/cockroachdb/pebble/internal/metamorphic/test.go:125 +0x788
        github.com/cockroachdb/pebble/internal/metamorphic.testMetaRun(0x14000069ce8?, {0x16b6cb815, 0x24}, 0x400?, {0x1400014c930, 0x2c})
                /Users/bilal/go/src/github.com/cockroachdb/pebble/internal/metamorphic/meta_test.go:202 +0x900
        github.com/cockroachdb/pebble/internal/metamorphic.TestMeta(0x1400022a9c0)
                /Users/bilal/go/src/github.com/cockroachdb/pebble/internal/metamorphic/meta_test.go:307 +0x6e8
        testing.tRunner(0x1400022a9c0, 0x104d910d8)
                /opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/testing/testing.go:1439 +0x110
        created by testing.(*T).Run
                /opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/testing/testing.go:1486 +0x300

        goroutine 20 [chan receive]:
        github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x104d974b0, 0x140001841b0})
                /Users/bilal/go/src/github.com/cockroachdb/pebble/table_cache.go:324 +0x84
        runtime/pprof.Do({0x104d97440?, 0x14000126008?}, {{0x14000160580?, 0x1400022a9c0?, 0x104834170?}}, 0x14000052f98)
                /opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/runtime/pprof/runtime.go:40 +0x84
        github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0x1400022a9c0?)
                /Users/bilal/go/src/github.com/cockroachdb/pebble/table_cache.go:322 +0x68
        created by github.com/cockroachdb/pebble.(*tableCacheShard).init
                /Users/bilal/go/src/github.com/cockroachdb/pebble/table_cache.go:314 +0xf8

        goroutine 21 [chan receive]:
        github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x104d974b0, 0x140004ac030})
                /Users/bilal/go/src/github.com/cockroachdb/pebble/table_cache.go:324 +0x84
        runtime/pprof.Do({0x104d97440?, 0x14000126008?}, {{0x14000160580?, 0x0?, 0x0?}}, 0x140004a8f98)
                /opt/homebrew/Cellar/go@1.18/1.18.5/libexec/src/runtime/pprof/runtime.go:40 +0x84
        github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0x0?)
 ```

**Comment by jbowens** [comment]
Nice!

Are we not actually scheduling a compaction earlier in the loop? Maybe because we abort for some other reason, or because we fail to find a file marked for compaction? Is the `curr.Stats.MarkedForCompaction` stat out of sync?

Or is there a race where the marked file is already being compacted? Oh maybe—`maybeScheduleCompactionPicker` can drop the mutex if another goroutine has the manifest lock. It'll `cond.Wait` waiting for the other goroutine to finish its manifest write. Once it's finished, we wake up and try to schedule the compaction for the file. If it was the last file marked for compaction, we don't schedule any compaction. Then we wait on `d.mu.compact.cond` with no other goroutine to wake us up.

**Comment by jbowens** [comment]
We should write a focused unit test that repros, but I think that makes as the culprit. Before we go to sleep on the condition variable we should assert that `d.mu.compact.compactingCount > 0`, because if it's not, something is amiss.

**Comment by itsbilal** [comment]
Still digging the code to come up with a theory that explains why we hang on that lock. Might be best to instrument the code or to add that assertion and then try stressing this run, yes.

For what's worth, this is the "history" of the run:

```
       ===== HISTORY =====
        // INFO: upgraded to format version: 005
        // INFO: upgraded to format version: 006
        // INFO: upgraded to format version: 007
        // INFO: upgraded to format version: 008
        // INFO: upgraded to format version: 009
        // INFO: [JOB 4] MANIFEST created 000226
        // INFO: upgraded to format version: 010
        // INFO: [JOB 6] compacting(rewrite) L5 [000225] (1.4 K) + L5 [] (0 B)
        // INFO: [JOB 7] compacting(rewrite) L6 [000224] (1.3 K) + L6 [] (0 B)
        // INFO: [JOB 6] compacting: sstable created 000228
        // INFO: [JOB 7] compacting: sstable created 000227
        // INFO: [JOB 6] compacted(rewrite) L5 [000225] (1.4 K) + L5 [] (0 B) -> L5 [000228] (1.7 K), in 0.0s (0.0s total), output rate 4.4 M/s
        // INFO: [JOB 6] sstable deleted 000225
        // INFO: [JOB 7] compacted(rewrite) L6 [000224] (1.3 K) + L6 [] (0 B) -> L6 [000227] (1.5 K), in 0.0s (0.0s total), output rate 6.8 M/s
        // INFO: [JOB 7] sstable deleted 000224
        // INFO: [JOB 9] all initial table stats loaded
```

I also just realized (and this will probably play a major role in the theory that we come up with): the hangs were all likely to happen on `standard-018`, which is the default options plus `max_concurrent_compactions = 2` (as opposed to the default of 1).

**Comment by itsbilal** [comment]
Ah, the answer was under my nose all this while. 

`maybeScheduleCompactionPicker` does a loop where it can schedule multiple compactions:

```
	for !d.opts.DisableAutomaticCompactions && d.mu.compact.compactingCount < maxConcurrentCompactions {
```

And here's the loop in `compactMarkedFilesLocked` (comments removed):

```
for curr.Stats.MarkedForCompaction > 0 {
		d.maybeScheduleCompactionPicker(func(picker compactionPicker, env compactionEnv) *pickedCompaction {
			return picker.pickRewriteCompaction(env)
		})

		if err := d.closed.Load(); err != nil {
			return err.(error)
		}
		d.mu.compact.cond.Wait()
		...
```

So the first time we call `maybeScheduleCompactionPicker`, we scheduled two compactions, one each for the two marked files. Then we waited once on the condition lock, got Broadcast()'d, looped and called `maybeScheduleCompactionPicker` again which scheduled nothing (because all candidate files are already compacting), but we waited again on the condition lock. And while we did the second call to `maybeScheduleCompactionPicker`, the table stats write was writing to the manifest so we had to drop `db.mu`, allowing the second compaction to finish. By the time we came back to the condition lock, all compactions were done.

I think the solution is to just not wait on the condition lock if `pickRewriteCompaction` returned `nil`. We can just capture that return value in the closure. An equivalent alternative is to not wait if `d.mu.compact.compactingCount == 0`.
